src/main/java/org/cmps/tetrahedron/utils/DataReader.java

```diff
@@ -19,15 +20,25 @@ public static Map<Integer, float[]> readVertices(File coordinatesTableFile) {
         try (Scanner fid = new Scanner(coordinatesTableFile)) {
             Map<Integer, float[]> coordinates = new HashMap<>();
 
-            while (fid.hasNext()) {
-                int index;
-                if (WITH_INDICES) {
-                    index = fid.nextInt();
+            while (fid.hasNextLine()) {
+                String[] elements = fid.nextLine().trim().split("\\s+");
+                int index, startIndex;
+                List<Float> vertex = new ArrayList<>();
+
+                if (elements.length == VERTICES_WITH_INDICES) {
+                    index = Integer.parseInt(elements[0]);
+                    startIndex = 1;
                 } else {
                     index = i++;
+                    startIndex = 0;
+                }
+
+                for (int j = startIndex; j < elements.length; j++) {
+                    vertex.add(Float.parseFloat(elements[j]));
                 }
 
-                coordinates.put(index, new float[]{fid.nextFloat(), fid.nextFloat(), fid.nextFloat()});
+                coordinates.put(index, convertListToArray(vertex));
```

Comment:
Can't we store coordinates in the array when we read them? This conversion looks redundant
We know the length of the coordinates array, so we shouldn't have any issues with creating it.